internal/clients/readiness.go

```diff
@@ -0,0 +1,146 @@
+package clients
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/adminapi"
+)
+
+const (
+	readinessCheckTimeout = time.Second
+)
+
+// ReadinessCheckResult represents the result of a readiness check.
+type ReadinessCheckResult struct {
+	// ClientsTurnedReady are the clients that were pending and are now ready to be used.
+	ClientsTurnedReady []*adminapi.Client
+	// ClientsTurnedPending are the clients that were ready and are now pending to be created.
+	ClientsTurnedPending []adminapi.DiscoveredAdminAPI
+}
+
+// HasChanges returns true if there are any changes in the readiness check result.
+// When no changes are present, it means that the readiness check haven't successfully created any pending client
+// nor detected any already created client that became not ready.
+func (r ReadinessCheckResult) HasChanges() bool {
+	return len(r.ClientsTurnedReady) > 0 || len(r.ClientsTurnedPending) > 0
+}
+
+// ReadinessChecker is responsible for checking the readiness of Admin API clients.
+type ReadinessChecker interface {
+	// CheckReadiness checks readiness of the provided clients:
+	// - alreadyCreatedClients are the clients that have already been created. The readiness of these clients will be
+	//   checked by their IsReady() method.
+	// - pendingClients are the clients that have not been created yet and are pending to be created. The readiness of
+	//   these clients will be checked by trying to create them.
+	CheckReadiness(
+		ctx context.Context,
+		alreadyCreatedClients []AlreadyCreatedClient,
+		pendingClients []adminapi.DiscoveredAdminAPI,
+	) ReadinessCheckResult
+}
+
+// AlreadyCreatedClient represents an Admin API client that has already been created.
+type AlreadyCreatedClient interface {
+	IsReady(context.Context) error
+	PodReference() (k8stypes.NamespacedName, bool)
+	BaseRootURL() string
+}
+
+type DefaultReadinessChecker struct {
+	factory ClientFactory
+	logger  logrus.FieldLogger
```

Comment:
Can we use `logr.Logger` here , not to increase the usage of `logrus.FieldLogger`?

Reply:
Updated it to use `logr.Logger` 👍